### ISSUE TITLE
cmd/go: go.work: go mod verify returns ziphash error in go.work

### ISSUE THREAD

**Issue by whs** [golang/go#54372]
### What version of Go are you using (`go version`)?

<pre>
$ go version
go version go1.19 linux/amd64
$ yay -Q go
go 2:1.19-1
</pre>

### Does this issue reproduce with the latest release?
Yes

### What operating system and processor architecture are you using (`go env`)?

<details><summary><code>go env</code> Output</summary><br><pre>
$ go env
GO111MODULE=""
GOARCH="amd64"
GOBIN=""
GOCACHE="/home/whs/.cache/go-build"
GOENV="/home/whs/.config/go/env"
GOEXE=""
GOEXPERIMENT=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOINSECURE=""
GOMODCACHE="/home/whs/build/go/pkg/mod"
GONOPROXY="git.wndv.co"
GONOSUMDB="git.wndv.co"
GOOS="linux"
GOPATH="/home/whs/build/go/"
GOPRIVATE=""
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/usr/lib/go"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/usr/lib/go/pkg/tool/linux_amd64"
GOVCS=""
GOVERSION="go1.19"
GCCGO="gccgo"
GOAMD64="v1"
AR="ar"
CC="gcc"
CXX="g++"
CGO_ENABLED="1"
GOMOD="/dev/null"
GOWORK="/home/whs/apps/gowork-repro/go.work"
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -Wl,--no-gc-sections -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build1279766172=/tmp/go-build -gno-record-gcc-switches"
</pre></details>

### What did you do?

1. Clone the [reproduction repository](https://github.com/whs/gowork-repro)
2. Run `go mod verify`

### What did you expect to see?
```
all modules verified
```

### What did you see instead?
```
github.com/whs/gowork-repro/mod2 : missing ziphash: open hash: no such file or directory
```

**Comment by thanm** [comment]
@golang/tools-team

**Comment by lopezator** [comment]
I fallen into the same scenario. Have to disable `go mod verify` in all the repos that use `go.work`

**Comment by sirianni** [comment]
+1 It's surprising to see no comments on this issue.  Is no one using workspaces or are we somehow doing it incorrectly?

**Comment by findleyr** [comment]
CC @bcmills @matloob

**Comment by findleyr** [comment]
CC @bcmills @matloob

**Comment by sirianni** [comment]
Related discussion on https://github.com/golang/go/issues/50750#issuecomment-1130020382

**Comment by jmwielandt** [comment]
![imagen](https://user-images.githubusercontent.com/37153467/204028281-25121e89-f9fd-456f-ae67-216d7df88777.png)
Not even setting GOPRIVATE to github.com (horrible practice, I'd expect to use github.com/<organization-name>) I could manage to make this work.

**Comment by bcmills** [comment]
`go mod verify` verifies modules that have been downloaded.

If the modules have not been downloaded (as might be the case when working exclusively within a larger workspace) it is supposed to skip modules that have not been downloaded. Probably something in that logic is not working as intended.

**Comment by freeformz** [comment]
FWIW: `go work sync` seems to usually clear this up for me.

**Comment by gabrielseibel1** [comment]
having the same issue, and go work sync does not solve it

**Comment by freeformz** [comment]
Actually nevermind. `go work sync` "solved" it once and only once.

**Comment by gbrlsnchs** [comment]
Is this really fixed? I can reproduce this by running `go mod verify` in a project like https://github.com/siderolabs/talos. I cloned it to `/tmp` and then got:

```console
$ go mod verify
module-sig-verify : missing ziphash: open hash: no such file or directory
```

```console
$ go version
go version go1.21.1 linux/amd64
```

```console
$ go env
GO111MODULE=''
GOARCH='amd64'
GOBIN=''
GOCACHE='/home/gbrlsnchs/.cache/go-build'
GOENV='/home/gbrlsnchs/.config/go/env'
GOEXE=''
GOEXPERIMENT=''
GOFLAGS=''
GOHOSTARCH='amd64'
GOHOSTOS='linux'
GOINSECURE=''
GOMODCACHE='/home/gbrlsnchs/go/pkg/mod'
GONOPROXY=''
GONOSUMDB=''
GOOS='linux'
GOPATH='/home/gbrlsnchs/go'
GOPRIVATE=''
GOPROXY='direct'
GOROOT='/usr/lib/go'
GOSUMDB='sum.golang.org'
GOTMPDIR=''
GOTOOLCHAIN='local'
GOTOOLDIR='/usr/lib/go/pkg/tool/linux_amd64'
GOVCS=''
GOVERSION='go1.21.1'
GCCGO='gccgo'
GOAMD64='v1'
AR='ar'
CC='x86_64-pc-linux-gnu-gcc'
CXX='x86_64-pc-linux-gnu-g++'
CGO_ENABLED='1'
GOMOD='/tmp/talos/go.mod'
GOWORK='/tmp/talos/go.work'
CGO_CFLAGS='-O2 -g'
CGO_CPPFLAGS=''
CGO_CXXFLAGS='-O2 -g'
CGO_FFLAGS='-O2 -g'
CGO_LDFLAGS='-O2 -g'
PKG_CONFIG='pkg-config'
GOGCCFLAGS='-fPIC -m64 -pthread -Wl,--no-gc-sections -fmessage-length=0 -ffile-prefix-map=/tmp/go-build975863248=/tmp/go-build -gno-record-gcc-switches'
```

**Comment by bcmills** [comment]
@gbrlsnchs, one underlying cause was fixed. If you are seeing a similar symptom, it likely has a different cause; please open a new issue with steps to reproduce the error. (Notably, it would help to have a specific commit at which you observed it.)
